generatorreceiver/internal/generator/trace_generator_test.go

```diff
@@ -0,0 +1,116 @@
+		"/product": {
+			DownstreamCalls: []topology.Call{
+				{Service: "productcatalogservice", Route: "/GetProducts"},
+				{Service: "recommendationservice", Route: "/GetRecommendations"},
+			},
+			MaxLatencyMillis: 10,
+		},
+	},
+}
+
+var topoTestFrontendWithoutDownStreamCalls = topology.ServiceTier{
+	Routes: map[string]*topology.ServiceRoute{
+		"/product": {MaxLatencyMillis: 10},
+	},
+}
+
+var topoTestCatalogService = topology.ServiceTier{
+	Routes: map[string]*topology.ServiceRoute{
+		"/GetProducts": {MaxLatencyMillis: 10},
+	},
+}
+
+var topoTestRecommendationService = topology.ServiceTier{
+	Routes: map[string]*topology.ServiceRoute{
+		"/GetRecommendations": {MaxLatencyMillis: 10},
+	},
+}
+
+func TestTraceGenerator_createSpanForServiceRouteCall(t *testing.T) {
+
+	type args struct {
+		traces         *ptrace.Traces
+		serviceName    string
+		routeName      string
+		startTimeNanos int64
+	}
+
+	traces := ptrace.NewTraces()
+	tests := []struct {
+		name string
+		topo *topology.Topology
+		args args
+	}{
+		{
+			name: "span with downstream calls",
+			topo: &topology.Topology{
+				Services: map[string]*topology.ServiceTier{
+					"frontend":              &topoTestFrontend,
+					"productcatalogservice": &topoTestCatalogService,
+					"recommendationservice": &topoTestRecommendationService,
+				},
+			},
+			args: args{
+				traces:         &traces,
+				serviceName:    "frontend",
+				routeName:      "/product",
+				startTimeNanos: 000000123,
+			},
+		},
+
+		{
+			name: "span without downstream calls",
+			topo: &topology.Topology{
+				Services: map[string]*topology.ServiceTier{
+					"frontend": &topoTestFrontendWithoutDownStreamCalls,
+				},
+			},
+			args: args{
+				traces:         &traces,
+				serviceName:    "frontend",
+				routeName:      "/product",
+				startTimeNanos: 000000123,
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewTraceGenerator(tt.topo, rand.New(rand.NewSource(rand.Int63())), tt.args.serviceName, tt.args.routeName)
+			genTraceID := g.genTraceId()
+			span := g.createSpanForServiceRouteCall(tt.args.traces, g.service, g.route, tt.args.startTimeNanos, genTraceID, pcommon.NewSpanIDEmpty())
+			require.Equal(t, span.Name(), tt.args.routeName)
+			require.Equal(t, span.TraceID(), genTraceID)
+			require.Equal(t, span.ParentSpanID(), pcommon.NewSpanIDEmpty()) //root span will have parent span id of 0
+
+			convertedSpanStartTime := pcommon.NewTimestampFromTime(time.Unix(0, tt.args.startTimeNanos))
+			require.Equal(t, span.StartTimestamp(), convertedSpanStartTime)
+
+			serviceTier := g.topology.GetServiceTier(tt.args.serviceName)
+			route := serviceTier.GetRoute(g.route)
+
+			for _, c := range route.DownstreamCalls {
+				var childStartTimeNanos = int64(span.StartTimestamp()) + route.SampleLatency(span.TraceID())
+				childSpan := g.createSpanForServiceRouteCall(tt.args.traces, c.Service, c.Route, childStartTimeNanos, span.TraceID(), g.genSpanId())
```

Comment:
Hm this doesn't feel right, since `createSpanForServiceRouteCall()` is a recursive function. So when we called it on line 97 for the root span, it also gets called for all the child spans. Will follow up with a more helpful comment..

Reply:
nice catch, are you thinking that I don't need to call it again here and I can straight up compare them against the input topology?

Reply:
So as we talked about, the child spans do in fact get stored in `resourceSpans` so I compared the contents of that to the topology from the test cases for tagSets and end times.